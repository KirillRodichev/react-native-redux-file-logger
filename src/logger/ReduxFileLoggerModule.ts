import { NativeModules, Platform } from 'react-native';
import type { FileConfig } from '../types';

const LINKING_ERROR =
  `The package 'react-native-redux-file-logger' doesn't seem to be linked. Make sure: \n\n` +
  Platform.select({ ios: "- You have run 'pod install'\n", default: '' }) +
  '- You rebuilt the app after installing the package\n' +
  '- You are not using Expo Go\n';

const ReduxFileLoggerModule = NativeModules.ReduxFileLogger
  ? NativeModules.ReduxFileLogger
  : new Proxy(
    {},
    {
      get() {
        throw new Error(LINKING_ERROR);
      },
    }
  );

type ReduxFileLoggerInterface = {
  addLogger: (tag: string, fileConfig: FileConfig) => Promise<void>;
  archive: (fileConfig: FileConfig, tag?: String) => Promise<string>;
  log: (tag: string, message: string) => void;
}

export default ReduxFileLoggerModule as ReduxFileLoggerInterface;
