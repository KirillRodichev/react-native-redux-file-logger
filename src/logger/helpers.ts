export function formatDate(date: Date | undefined): string | undefined {
    return date && `${date.toLocaleString()}.${date.getMilliseconds()}`;
}

export function getDuration(start: Date | undefined, end: Date | undefined): string | undefined {
    return start && end && `${end.getMilliseconds() - start.getMilliseconds()} ms`;
}
