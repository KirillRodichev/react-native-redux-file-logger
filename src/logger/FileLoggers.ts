import ReduxFileLoggerModule from './ReduxFileLoggerModule';
import type { FileConfig, Logger } from '../types';

export class FileLoggers {
  private readonly fileLoggers = new Map<string, Logger>();

  public addLogger = async (tag: string, fileConfig: FileConfig) => {
    await ReduxFileLoggerModule.addLogger(tag, fileConfig);
    this.fileLoggers.set(tag, {
      log: message => ReduxFileLoggerModule.log(tag, message),
    })
  }

  public getLogger = (tag: string) => {
    return this.fileLoggers.get(tag);
  }
}
