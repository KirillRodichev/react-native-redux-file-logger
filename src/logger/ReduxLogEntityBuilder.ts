import type { AnyAction } from 'redux';
import type { LoggerOptions } from 'react-native-redux-file-logger';
import { formatDate, getDuration } from './helpers';

export type Requirements = Pick<LoggerOptions, 'showDiff' | 'shouldLogPrevState' | 'shouldLogNextState'>;

export class ReduxLogEntityBuilder<TState = any> {
    private readonly action: AnyAction;
    private diff?: any;
    private prevState?: TState;
    private nextState?: TState;
    private startTime?: Date;
    private endTime?: Date;
    private requirements: Requirements;

    constructor(action: AnyAction, requirements: Requirements) {
        this.action = action;
        this.requirements = requirements;
    }

    public addPrevState(prevState: TState) {
        if (this.requirements.shouldLogPrevState) {
            this.prevState = prevState;
        }
        return this;
    }

    public addNextState(nextState: TState) {
        if (this.requirements.shouldLogNextState) {
            this.nextState = nextState;
        }
        return this;
    }

    public addDiff(diff: any) {
        if (this.requirements.showDiff) {
            this.diff = diff;
        }
        return this;
    }

    public addStartTime(date: Date) {
        this.startTime = date;
        return this;
    }

    public addEndTime(date: Date) {
        this.endTime = date;
        return this;
    }

    public build(): string {
        return this.stringify();
    }

    private stringify() {
        const formattedStartDate = formatDate(this.startTime);
        const duration = getDuration(this.startTime, this.endTime);
        const key = `action: ${this.action.type} @ ${formattedStartDate} (in ${duration})`;

        const logEntity = {
            action: this.action,
            prevState: this.prevState,
            diff: this.diff,
            nextState: this.nextState,
        };

        return `"${key}": ${JSON.stringify(logEntity)},`;
    }
}
