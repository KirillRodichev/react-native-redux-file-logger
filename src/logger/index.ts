import { FileLoggers } from './FileLoggers';

export const { addLogger: addFileLogger, getLogger: getFileLogger } = new FileLoggers();
