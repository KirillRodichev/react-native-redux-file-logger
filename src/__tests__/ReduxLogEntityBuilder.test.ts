import type { AnyAction } from 'redux';
import { diff } from 'deep-object-diff';
import { ReduxLogEntityBuilder, Requirements } from '../logger/ReduxLogEntityBuilder';

let logEntryBuilder: ReduxLogEntityBuilder;
const actionObj: AnyAction = { type: 'testType' };
const requirements: Requirements = {
    showDiff: true,
    shouldLogNextState: true,
    shouldLogPrevState: true,
};

beforeAll(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date(2022, 11, 1));
});

afterAll(() => {
    jest.useRealTimers();
});

beforeEach(() => {
    logEntryBuilder = new ReduxLogEntityBuilder(actionObj, requirements);
    logEntryBuilder.addStartTime(new Date()).addEndTime(new Date());
});

describe('ReduxLogEntityBuilder', () => {
    const prevState = { counter: 1 };
    const nextState = { counter: 2 };

    describe('with all requirements', () => {
        it('should build with prev state', () => {
            logEntryBuilder.addPrevState(prevState);

            expect(logEntryBuilder.build()).toMatchSnapshot();
        });

        it('should build with next state', () => {
            logEntryBuilder.addNextState(nextState);

            expect(logEntryBuilder.build()).toMatchSnapshot();
        });

        it('should build with diff', () => {
            logEntryBuilder.addDiff(diff(prevState as object, nextState as object));

            expect(logEntryBuilder.build()).toMatchSnapshot();
        });

        it('should build all requirements', () => {
            logEntryBuilder.addPrevState(prevState);
            logEntryBuilder.addNextState(nextState);
            logEntryBuilder.addDiff(diff(prevState as object, nextState as object));

            expect(logEntryBuilder.build()).toMatchSnapshot();
        });
    });

    describe('without requirements', () => {
        it('should build empty action', () => {
            const emptyRequirements = {};
            const emptyActionBuilder = new ReduxLogEntityBuilder(actionObj, emptyRequirements);

            emptyActionBuilder.addStartTime(new Date()).addPrevState(prevState);
            emptyActionBuilder.addEndTime(new Date()).addNextState(nextState);
            emptyActionBuilder.addDiff(diff(prevState as object, nextState as object));

            expect(emptyActionBuilder.build()).toMatchSnapshot();
        });
    });
});
