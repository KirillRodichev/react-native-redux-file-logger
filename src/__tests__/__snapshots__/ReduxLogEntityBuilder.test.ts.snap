// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReduxLogEntityBuilder with all requirements should build all requirements 1`] = `"\\"action: testType @ 12/1/2022, 12:00:00 AM.0 (in 0 ms)\\": {\\"action\\":{\\"type\\":\\"testType\\"},\\"prevState\\":{\\"counter\\":1},\\"diff\\":{\\"counter\\":2},\\"nextState\\":{\\"counter\\":2}},"`;

exports[`ReduxLogEntityBuilder with all requirements should build with diff 1`] = `"\\"action: testType @ 12/1/2022, 12:00:00 AM.0 (in 0 ms)\\": {\\"action\\":{\\"type\\":\\"testType\\"},\\"diff\\":{\\"counter\\":2}},"`;

exports[`ReduxLogEntityBuilder with all requirements should build with next state 1`] = `"\\"action: testType @ 12/1/2022, 12:00:00 AM.0 (in 0 ms)\\": {\\"action\\":{\\"type\\":\\"testType\\"},\\"nextState\\":{\\"counter\\":2}},"`;

exports[`ReduxLogEntityBuilder with all requirements should build with prev state 1`] = `"\\"action: testType @ 12/1/2022, 12:00:00 AM.0 (in 0 ms)\\": {\\"action\\":{\\"type\\":\\"testType\\"},\\"prevState\\":{\\"counter\\":1}},"`;

exports[`ReduxLogEntityBuilder without requirements should build empty action 1`] = `"\\"action: testType @ 12/1/2022, 12:00:00 AM.0 (in 0 ms)\\": {\\"action\\":{\\"type\\":\\"testType\\"}},"`;
